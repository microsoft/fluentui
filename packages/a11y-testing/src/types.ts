// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type PropValue = string | number | boolean | any;
export type Props = { [name: string]: PropValue };
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AccessibilityBehavior = (props: Props) => any;

export interface TestFacade {
  slotExists(slotName: string): boolean;
  attributeExists(slotName: string, attributeName: string): boolean;
  attributeHasValue(slotName: string, attributeName: string, value: PropValue): boolean;
  getAttributeValue(slotName: string, attribute: string): PropValue;
  forProps(props: Props): TestFacade;
  afterClick(slotName: string): void;
  pressSpaceKey(slotName: string): void;
  pressEnterKey(slotName: string): void;
  verifyOnclickExecution(slotName: string): boolean;
}

export interface Rule {
  stringify: () => string;
}

export interface Slot {
  name: string;
  props: Props[];
  expectAttribute: boolean;
  expectedAttribute: string;
  expectedValue: PropValue;
  description: string;
  hidden: boolean;
  afterEventData: Event;
  checkClick: boolean;
  checkSpaceKeyPressed: boolean;
  checkEnterKeyPressed: boolean;
  checkOnClickWasExecuted: boolean;
  overrideId: boolean;
}
