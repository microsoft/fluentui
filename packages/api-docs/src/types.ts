/**
 * Options that define the information necessary to find the api.json file
 * and create individual page json files out of it
 */
export interface IPageJsonOptions {
  /** Input files generated by api-extractor (one per package) */
  apiJsonPaths: string[];
  /** Output \*.page.json folder path. Page groups (if used) will go in subfolders. */
  outputRoot: string;
  /** If true, the generated files won't be pretty-formatted */
  min?: boolean;
  /**
   * List of allowed page names (`@docCategory` tag values) grouped by any desired criteria,
   * such as package name.
   */
  pageGroups?: PageGroups;
  /**
   * Group for any page names (`@docCategory` tag values) not listed in `pageGroups`.
   * If unspecified, unlisted pages will go under the root.
   */
  fallbackGroup?: string;
}

/** Map from group name to page names */
export type PageGroups = { [groupName: string]: string[] };

/**
 * Structure of the page.json files
 */
export interface IPageJson {
  tables: ITableJson[];
  name: string;
  group?: string;
}

export type ApiKind = 'interface' | 'enum' | 'class' | 'typeAlias';

/**
 * Info for a table representing a top-level API item: interface, enum, class, or type alias.
 */
export interface ITableJson {
  kind: ApiKind;
  name: string;
  description?: string;
  /**
   * Any types the item extends, translated to an array of text elements and links to other types.
   * For classes and interfaces only.
   *
   * Example: `Readonly<IFoo>` might translate to:
   * `[{ text: 'Readonly<' }, { text: 'IFoo', linkedPage: 'Foo', linkedPageGroup: 'components' }, { text: '>' }]`
   */
  extendsTokens?: ILinkToken[];
  members?: ITableRowJson[] | IEnumTableRowJson[];
  deprecated?: boolean;
  deprecatedMessage?: string;
}

/**
 * Generic row for API reference tables.
 * It can represent a member (property or method) of an interface or class.
 */
export interface ITableRowJson {
  name: string;
  kind?: 'method' | 'property';
  /**
   * The row's type translated to an array of text elements and links to other types.
   * For example, `Readonly<IFoo>` might translate to:
   * `[{ text: 'Readonly<' }, { text: 'IFoo', linkedPage: 'Foo', linkedPageGroup: 'components' }, { text: '>' }]`
   */
  typeTokens: ILinkToken[];
  defaultValue?: string;
  description?: string;
  deprecated?: boolean;
  deprecatedMessage?: string;
  required?: boolean;
}

/**
 * Enum member row for API reference tables.
 */
export type IEnumTableRowJson = Omit<ITableRowJson, 'kind' | 'typeTokens' | 'defaultValue' | 'required'> & {
  value: string;
};

/**
 * Text excerpt token that is part of a type definition or extends block and may have a link
 * to another doc page.
 */
export interface ILinkToken {
  text: string;
  /** If this token is a link, name of the doc page it points to */
  linkedPage?: string;
  /** If this token is a link, group/category of the doc page it points to */
  linkedPageGroup?: string;
}
