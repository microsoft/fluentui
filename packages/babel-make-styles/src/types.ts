import type { TransformOptions } from '@babel/core';
import type { EvalRule } from '@linaria/babel-preset';

export type BabelPluginOptions = {
  /** Defines set of modules and imports handled by a plugin. */
  modules?: { moduleSource: string; importName: string }[];

  /**
   * If you need to specify custom Babel configuration, you can pass them here. These options will be used by the
   * plugin when parsing and evaluating modules.
   */
  babelOptions?: Pick<TransformOptions, 'plugins' | 'presets'>;

  /** The set of rules that defines how the matched files will be transformed during the evaluation. */
  evaluationRules?: EvalRule[];
};
