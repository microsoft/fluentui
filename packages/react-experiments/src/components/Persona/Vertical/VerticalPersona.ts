import * as React from 'react';
import { VerticalPersonaView } from './VerticalPersona.view';
import { VerticalPersonaStyles, VerticalPersonaTokens } from './VerticalPersona.styles';
import { createComponent } from '@fluentui/foundation-legacy';
import type { IVerticalPersonaProps } from './VerticalPersona.types';

export const VerticalPersona: React.FunctionComponent<IVerticalPersonaProps> = createComponent(VerticalPersonaView, {
  displayName: 'VerticalPersona',
  styles: VerticalPersonaStyles,
  tokens: VerticalPersonaTokens,
});
