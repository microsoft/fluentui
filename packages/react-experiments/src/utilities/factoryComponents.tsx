import * as React from 'react';
import {
  FontIcon as FabricFontIcon,
  Icon as FabricIcon,
  Label as FabricLabel,
  IFontIconProps,
  IIconProps,
  ILabelProps,
  IPersonaPresenceProps,
} from '@fluentui/react';
// PersonaPresence is not exported by OUFR, so we have to import it directly.
import { PersonaPresence as FabricPersonaPresence } from '@fluentui/react/lib/PersonaPresence';
import { createFactory, ISlottableComponentType, ISlotFactory } from '@fluentui/foundation-legacy';

// TODO: All contents of this file should be moved to each respective component as they are converted to use slots.
// TODO: createFactory should no longer have to be explicitly called with component options containing defaultProp.
//       (Consider adding a defaultProp option to styled so that createFactory can be internalized similar
//       to createComponent)
// TODO: displayName will also be covered by createComponent argument.
// TODO: These components will also have to be modified not to generate rendered output if their props don't call for it
//       to eliminate the wrapper functions below.

// Generally to avoid a bunch of "if slot prop exists" checks in parent components, components should
// make sure they have content to render based on their props. For example here, if Icon has no iconName,
// it has no rendered content and returns null. This prevents Button.view from having to check to
// see if its icon Slot is defined.

// These wrappers will temporarily add a layer to the hierarchy (identified with displayName) until their functionality
// can be absorbed into their respective OUFR components.
export const FontIcon: ISlottableComponentType<IFontIconProps, string> = props =>
  props.iconName ? <FabricFontIcon {...props} /> : null;
FontIcon.displayName = 'FontIcon';
const fontIconFactory: ISlotFactory<IFontIconProps, string> = createFactory(FontIcon, { defaultProp: 'iconName' });
FontIcon.create = fontIconFactory;

export const Icon: ISlottableComponentType<IIconProps, string> = props =>
  props.iconName ? <FabricIcon {...props} /> : null;
Icon.displayName = 'Icon';
const iconFactory: ISlotFactory<IIconProps, string> = createFactory(Icon, { defaultProp: 'iconName' });
Icon.create = iconFactory;

export const Label: ISlottableComponentType<ILabelProps, string> = props =>
  React.Children.count(props.children) > 0 ? <FabricLabel {...props} /> : null;
Label.displayName = 'Label';
const labelFactory = createFactory(Label) as ISlotFactory<ILabelProps, string>;
Label.create = labelFactory;

export const PersonaPresence: ISlottableComponentType<IPersonaPresenceProps, number> =
  // TODO: This is a bug in PersonaPresence that needs to be fixed. 'presence' prop comment mentions that it
  // won't render if presence is undefined, but it does render. Check for undefined here for now.
  props => (props.presence !== undefined ? <FabricPersonaPresence {...props} /> : null);
PersonaPresence.displayName = 'PersonaPresence';
const personaPresenceFactory: ISlotFactory<IPersonaPresenceProps, number> = createFactory(PersonaPresence, {
  defaultProp: 'presence',
});
PersonaPresence.create = personaPresenceFactory;
