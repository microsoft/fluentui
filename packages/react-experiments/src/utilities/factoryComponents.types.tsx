import {
  IButtonProps,
  ICalloutProps,
  IContextualMenuProps,
  IFontIconProps,
  IIconProps,
  ILabelProps,
  IListProps,
  IPersonaPresenceProps,
  PersonaPresence,
} from '@fluentui/react';
import { ISlotProp } from '@fluentui/foundation-legacy';

// TODO: All contents of this file should be moved to each respective component after slots utilities are promoted.

export type IButtonSlot = ISlotProp<IButtonProps>;
export type ICalloutSlot = ISlotProp<ICalloutProps>;
export type IContextualMenuSlot = ISlotProp<IContextualMenuProps>;
export type IFontIconSlot = ISlotProp<IFontIconProps, string>;
export type IIconSlot = ISlotProp<IIconProps, string>;
export type ILabelSlot = ISlotProp<ILabelProps, string>;
export type IListSlot = ISlotProp<IListProps>;
export type IPersonaPresenceSlot = ISlotProp<IPersonaPresenceProps, PersonaPresence>;
