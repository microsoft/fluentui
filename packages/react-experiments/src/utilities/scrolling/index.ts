export * from './ScrollContainer.types';
export * from './ScrollContainer';
