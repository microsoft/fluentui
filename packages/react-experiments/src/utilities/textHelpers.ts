import { memoizeFunction } from '@fluentui/react/lib/Utilities';

const getContextOrFalse = memoizeFunction(
  (): CanvasRenderingContext2D | false => document.createElement('canvas').getContext('2d') || false,
  1,
);

export function isSingleLineText(lineLength: number, text: string, textStyle: string): boolean {
  const context = getContextOrFalse();

  if (!context) {
    return false;
  }

  context.font = textStyle;

  const textWidth = context.measureText(text).width;

  return textWidth <= lineLength;
}
