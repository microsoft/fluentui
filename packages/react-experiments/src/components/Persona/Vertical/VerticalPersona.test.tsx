import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as TextHelpers from '../../../utilities/textHelpers';

import { IVerticalPersonaComponent } from './VerticalPersona.types';
import { VerticalPersona } from './VerticalPersona';

const testVerticalPersonaStyles: IVerticalPersonaComponent['styles'] = {
  root: 'test-cn-root',
  primaryText: 'test-cn-text',
  secondaryText: 'test-cn-secondaryText',
  coin: 'test-cn-coin',
};

// Views are just pure functions with no statefulness, which means they can get full code coverage
// with snapshot tests exercising permutations of the props.
describe('VerticalPersona', () => {
  beforeAll(() => {
    // Prevent canvas usage from failing test
    jest.spyOn(TextHelpers, 'isSingleLineText').mockImplementation((() => {
      /** no impl **/
    }) as any);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('renders correctly with only a text', () => {
    const tree = renderer
      .create(<VerticalPersona vertical text="James Bond" styles={testVerticalPersonaStyles} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a text and secondary text', () => {
    const tree = renderer
      .create(
        <VerticalPersona
          vertical
          text="James Bond"
          secondaryText="Super secret agent"
          styles={testVerticalPersonaStyles}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the coin with the passed coinProps', () => {
    const tree = renderer
      .create(
        <VerticalPersona
          vertical
          text="James Bond"
          secondaryText="Super secret agent"
          styles={testVerticalPersonaStyles}
          coin={{ initials: 'MI6', coinColor: 'red' }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
