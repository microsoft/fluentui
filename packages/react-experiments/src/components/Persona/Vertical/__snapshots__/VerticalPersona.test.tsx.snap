// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalPersona renders correctly with a text and secondary text 1`] = `
<div
  className=
      test-cn-root
      {
        align-items: center;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        justify-content: flex-start;
        padding-bottom: 0;
        padding-left: 8px;
        padding-right: 8px;
        padding-top: 0;
        width: 122px;
      }
>
  <div
    className=

        {
          align-items: center;
          background-color: #A4262C;
          border-radius: 50%;
          color: white;
          display: flex;
          height: 48px;
          justify-content: center;
          position: relative;
          width: 48px;
        }
  >
    <span
      className=

          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            color: #323130;
            display: inline;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 19.2px;
            font-weight: 400;
            moz-osx-font-smoothing: grayscale;
            webkit-font-smoothing: antialiased;
          }
    >
      JB
    </span>
  </div>
  <span
    className=
        test-cn-text
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #323130;
          display: inline;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 600;
          moz-osx-font-smoothing: grayscale;
          padding-top: 8px;
          text-align: center;
          webkit-font-smoothing: antialiased;
          white-space: initial;
          width: 100%;
          word-break: break-word;
          word-wrap: break-word;
        }
  >
    James Bond
  </span>
  <span
    className=
        test-cn-secondaryText
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #605e5c;
          display: inline;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 12px;
          font-weight: 400;
          line-height: 1.42;
          moz-osx-font-smoothing: grayscale;
          padding-top: 6px;
          text-align: center;
          webkit-font-smoothing: antialiased;
          white-space: initial;
          width: 100%;
          word-break: break-word;
          word-wrap: break-word;
        }
  >
    Super secret agent
  </span>
</div>
`;

exports[`VerticalPersona renders correctly with only a text 1`] = `
<div
  className=
      test-cn-root
      {
        align-items: center;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        justify-content: flex-start;
        padding-bottom: 0;
        padding-left: 8px;
        padding-right: 8px;
        padding-top: 0;
        width: 122px;
      }
>
  <div
    className=

        {
          align-items: center;
          background-color: #A4262C;
          border-radius: 50%;
          color: white;
          display: flex;
          height: 48px;
          justify-content: center;
          position: relative;
          width: 48px;
        }
  >
    <span
      className=

          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            color: #323130;
            display: inline;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 19.2px;
            font-weight: 400;
            moz-osx-font-smoothing: grayscale;
            webkit-font-smoothing: antialiased;
          }
    >
      JB
    </span>
  </div>
  <span
    className=
        test-cn-text
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #323130;
          display: inline;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 600;
          moz-osx-font-smoothing: grayscale;
          padding-top: 8px;
          text-align: center;
          webkit-font-smoothing: antialiased;
          white-space: initial;
          width: 100%;
          word-break: break-word;
          word-wrap: break-word;
        }
  >
    James Bond
  </span>
</div>
`;

exports[`VerticalPersona renders the coin with the passed coinProps 1`] = `
<div
  className=
      test-cn-root
      {
        align-items: center;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        justify-content: flex-start;
        padding-bottom: 0;
        padding-left: 8px;
        padding-right: 8px;
        padding-top: 0;
        width: 122px;
      }
>
  <div
    className=

        {
          align-items: center;
          background-color: red;
          border-radius: 50%;
          color: white;
          display: flex;
          height: 48px;
          justify-content: center;
          position: relative;
          width: 48px;
        }
  >
    <span
      className=

          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            color: #323130;
            display: inline;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 19.2px;
            font-weight: 400;
            moz-osx-font-smoothing: grayscale;
            webkit-font-smoothing: antialiased;
          }
    >
      MI6
    </span>
  </div>
  <span
    className=
        test-cn-text
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #323130;
          display: inline;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 600;
          moz-osx-font-smoothing: grayscale;
          padding-top: 8px;
          text-align: center;
          webkit-font-smoothing: antialiased;
          white-space: initial;
          width: 100%;
          word-break: break-word;
          word-wrap: break-word;
        }
  >
    James Bond
  </span>
  <span
    className=
        test-cn-secondaryText
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #605e5c;
          display: inline;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 12px;
          font-weight: 400;
          line-height: 1.42;
          moz-osx-font-smoothing: grayscale;
          padding-top: 6px;
          text-align: center;
          webkit-font-smoothing: antialiased;
          white-space: initial;
          width: 100%;
          word-break: break-word;
          word-wrap: break-word;
        }
  >
    Super secret agent
  </span>
</div>
`;
